server/sonar-server/src/main/java/org/sonar/server/permission/ws/PermissionWsSupport.java

```diff
@@ -0,0 +1,105 @@
+import org.sonar.server.permission.ProjectId;
+import org.sonar.server.permission.UserId;
+import org.sonar.server.permission.ws.template.WsTemplateRef;
+import org.sonar.server.usergroups.ws.GroupIdOrAnyone;
+import org.sonar.server.usergroups.ws.GroupWsRef;
+import org.sonar.server.usergroups.ws.GroupWsSupport;
+import org.sonarqube.ws.client.permission.PermissionsWsParameters;
+
+import static org.sonar.server.ws.WsUtils.checkFound;
+import static org.sonarqube.ws.client.permission.PermissionsWsParameters.PARAM_GROUP_ID;
+import static org.sonarqube.ws.client.permission.PermissionsWsParameters.PARAM_GROUP_NAME;
+import static org.sonarqube.ws.client.permission.PermissionsWsParameters.PARAM_ORGANIZATION_KEY;
+
+public class PermissionWsSupport {
+
+  private final DbClient dbClient;
+  private final ComponentFinder componentFinder;
+  private final GroupWsSupport groupWsSupport;
+  private final ResourceTypes resourceTypes;
+
+  public PermissionWsSupport(DbClient dbClient, ComponentFinder componentFinder, GroupWsSupport groupWsSupport, ResourceTypes resourceTypes) {
+    this.dbClient = dbClient;
+    this.componentFinder = componentFinder;
+    this.groupWsSupport = groupWsSupport;
+    this.resourceTypes = resourceTypes;
+  }
+
+  /**
+   * @throws org.sonar.server.exceptions.NotFoundException if a project does not exist
+   */
+  public ProjectId findProject(DbSession dbSession, WsProjectRef ref) {
+    ComponentDto project = componentFinder.getRootComponentOrModuleByUuidOrKey(dbSession, ref.uuid(), ref.key(), resourceTypes);
+    return new ProjectId(project.getId(), project.uuid());
+  }
+
+  public Optional<ProjectId> findProject(DbSession dbSession, Request request) {
+    String uuid = request.param(PermissionsWsParameters.PARAM_PROJECT_ID);
+    String key = request.param(PermissionsWsParameters.PARAM_PROJECT_KEY);
+    if (uuid != null || key != null) {
+      WsProjectRef ref = WsProjectRef.newWsProjectRef(uuid, key);
+      return Optional.of(findProject(dbSession, ref));
+    }
+    return Optional.empty();
+  }
+
+  public ComponentDto getRootComponentOrModule(DbSession dbSession, WsProjectRef projectRef) {
+    return componentFinder.getRootComponentOrModuleByUuidOrKey(dbSession, projectRef.uuid(), projectRef.key(), resourceTypes);
+  }
+
+  public GroupIdOrAnyone findGroup(DbSession dbSession, Request request) {
+    Long groupId = request.paramAsLong(PARAM_GROUP_ID);
+    String orgKey = request.param(PARAM_ORGANIZATION_KEY);
+    String groupName = request.param(PARAM_GROUP_NAME);
+    GroupWsRef groupRef = groupWsSupport.newGroupRef(dbSession, groupId, orgKey, groupName);
+    return groupWsSupport.findGroupOrAnyone(dbSession, groupRef);
+  }
+
+  public UserId findUser(DbSession dbSession, String login) {
+    UserDto dto = dbClient.userDao().selectActiveUserByLogin(dbSession, login);
+    checkFound(dto, "User with login '%s' is not found'", login);
+    return new UserId(dto.getId(), dto.getLogin());
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) NullPointerException might be thrown as 'dto' is nullable here [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2259)